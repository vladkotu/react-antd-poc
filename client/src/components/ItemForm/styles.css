.ItemForm-layout .ant-form-item{
  margin-bottom: 10px;
}

.ItemForm-layout .ant-form-item .ant-form-item-label{
  padding-bottom: 0px;
}

.ItemForm-layout .ant-btn{
  margin-right: 8px;
}
