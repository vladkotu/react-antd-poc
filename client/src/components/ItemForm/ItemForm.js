import React from 'react'
import PropTypes from 'prop-types'
import { Form, InputNumber, Button } from 'antd'
import './styles.css'

const layout = {
  labelCol: { xs: { span: 16 }, sm: { span: 8 } },
  wrapperCol: { span: 12 },
}

const tailLayout = {
  wrapperCol: { xs: { span: 16, offset: 0 }, sm: { span: 16, offset: 8 } },
}

function ItemForm({
  name,
  currentItem = {},
  fields,
  onCancel,
  onFinish,
  onRemove,
}) {
  const formName = name ? name : `item-form-${currentItem.id || 'no-name'}`
  return (
    <Form
      className='ItemForm-layout'
      {...layout}
      name={formName}
      initialValues={currentItem}
      onFinish={onFinish}
    >
      <Form.Item name='id' style={{ display: 'none' }}>
        <InputNumber />
      </Form.Item>

      {fields.map(({ field, ...fieldProps }) => (
        <Form.Item key={fieldProps.name} {...fieldProps}>
          {field}
        </Form.Item>
      ))}

      <Form.Item {...tailLayout}>
        <Button type='primary' htmlType='submit'>
          Save Changes
        </Button>
        <Button onClick={onCancel}>Cancel</Button>
        {onRemove && currentItem && (
          <Button onClick={() => onRemove(currentItem)}>Remove</Button>
        )}
      </Form.Item>
    </Form>
  )
}

ItemForm.propTypes = {
  name: PropTypes.string,
  currentItem: PropTypes.object,
  fields: PropTypes.array,
  onCancel: PropTypes.func,
  onFinish: PropTypes.func,
  onRemove: PropTypes.func,
}

export default ItemForm
