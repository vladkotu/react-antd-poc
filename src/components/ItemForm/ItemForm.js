import React from 'react'
import { Form, InputNumber, Button } from 'antd'
import './styles.css'

const layout = {
  labelCol: { xs: { span: 16 }, sm: { span: 8 } },
  wrapperCol: { span: 16 },
}

const tailLayout = {
  wrapperCol: { xs: { span: 16, offset: 0 }, sm: { span: 24, offset: 0 } },
}

function ItemForm({
  name,
  currentItem = {},
  fields,
  onCancel,
  onFinish,
  onRemove,
}) {
  const formName = name ? name : `item-form-${currentItem.id || 'no-name'}`
  return (
    <Form
      className='ItemForm-layout'
      {...layout}
      name={formName}
      initialValues={currentItem}
      onFinish={onFinish}
    >
      <Form.Item name='id' style={{ display: 'none' }}>
        <InputNumber />
      </Form.Item>

      {fields.map(({ field, ...fieldProps }) => (
        <Form.Item key={fieldProps.name} {...fieldProps}>
          {field}
        </Form.Item>
      ))}

      <Form.Item {...tailLayout}>
        <Button type='primary' htmlType='submit'>
          Save Changes
        </Button>
        <Button onClick={onCancel}>Cancel</Button>
        {onRemove && currentItem && (
          <Button onClick={() => onRemove(currentItem)}>Remove</Button>
        )}
      </Form.Item>
    </Form>
  )
}

export default ItemForm
