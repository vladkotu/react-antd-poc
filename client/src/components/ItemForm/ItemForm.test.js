import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Form, InputNumber, Input } from 'antd'
import ItemForm from './ItemForm'

function makeProps(props) {
  return {
    name: 'testForm',
    fields: [],
    currentItem: { id: 'testId' },
    onRemove: jest.fn(),
    onCancel: jest.fn(),
    onFinish: jest.fn(),
    ...props,
  }
}

describe('ItemForm', () => {
  describe('default fields', () => {
    let utils = {}
    let props = {}
    beforeEach(() => {
      props = makeProps({ name: 'empty-test-form' })
      utils = render(<ItemForm {...props} />)
    })

    it('renders hidden id input field that stores item id', () => {
      const { container } = utils
      const hiddenInput = container.querySelectorAll('input')
      expect(hiddenInput.length).toBe(1)
      expect(hiddenInput[0]).toHaveAttribute('id', `${props.name}_id`)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('renders cancel action control', () => {
      const { getByText } = utils
      const button = getByText(/Cancel/i)
      fireEvent.click(button, { button: 1 })
      expect(props.onCancel).toHaveBeenCalled()
      expect(button).toBeInTheDocument()
    })

    it('renders remove action control conditionally', () => {
      const { getByText } = utils
      const button = getByText(/Remove/i)
      fireEvent.click(button, { button: 1 })
      expect(props.onRemove).toHaveBeenCalled()
      expect(button).toBeInTheDocument()
    })
  })

  describe('renders custom fields', () => {
    let utils = {}
    let props = {}
    beforeEach(() => {
      props = makeProps({
        name: 'custom-test-form',
        fields: [
          {
            name: 'numfield',
            label: 'Number Field',
            rules: [
              {
                required: true,
              },
            ],
            field: <InputNumber />,
          },
          {
            name: 'text-field',
            label: 'Text Field',
            field: <Input />,
          },
        ],
      })
      utils = render(<ItemForm {...props} />)
    })

    it('renders 2 input fields', () => {
      const { getByLabelText } = utils
      const field1 = getByLabelText(/Number Field/i)
      const field2 = getByLabelText(/Text Field/i)
      expect(field1).toBeInTheDocument()
      expect(field2).toBeInTheDocument()
    })

    const delay = (timeout = 0) =>
      new Promise(resolve => {
        setTimeout(resolve, timeout)
      })

    it('does not submits values if required field is not filled', async () => {
      const { container } = utils
      const form = container.querySelector('form')
      fireEvent.submit(form)
      await delay()
      expect(props.onFinish).not.toHaveBeenCalled()
    })

    it('submits proper values', async () => {
      const { getByLabelText, container } = utils
      const field1 = getByLabelText(/Number Field/i)
      const field2 = getByLabelText(/Text Field/i)
      const form = container.querySelector('form')
      fireEvent.change(field1, { target: { value: 3 } })
      fireEvent.change(field2, { target: { value: 'field value' } })
      fireEvent.submit(form)
      await delay()
      expect(props.onFinish).toHaveBeenCalledWith({
        id: props.currentItem.id,
        numfield: 3,
        'text-field': 'field value',
      })
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
