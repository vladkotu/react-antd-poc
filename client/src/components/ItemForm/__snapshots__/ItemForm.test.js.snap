// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemForm default fields renders hidden id input field that stores item id 1`] = `
<form
  class="ant-form ant-form-horizontal ItemForm-layout"
  id="empty-test-form"
>
  <div
    class="ant-row ant-form-item"
    style="display: none;"
  >
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-input-number"
          >
            <div
              class="ant-input-number-handler-wrap"
            >
              <span
                aria-disabled="true"
                aria-label="Increase Value"
                class="ant-input-number-handler ant-input-number-handler-up ant-input-number-handler-up-disabled"
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="up"
                  class="anticon anticon-up ant-input-number-handler-up-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="up"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                    />
                  </svg>
                </span>
              </span>
              <span
                aria-disabled="true"
                aria-label="Decrease Value"
                class="ant-input-number-handler ant-input-number-handler-down ant-input-number-handler-down-disabled"
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-input-number-handler-down-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="ant-input-number-input-wrap"
            >
              <input
                aria-valuemin="-9007199254740991"
                aria-valuenow="testId"
                autocomplete="off"
                class="ant-input-number-input"
                id="empty-test-form_id"
                min="-9007199254740991"
                role="spinbutton"
                step="1"
                value="testId"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-control ant-col-xs-16 ant-col-xs-offset-0 ant-col-sm-16 ant-col-sm-offset-8"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <button
            class="ant-btn ant-btn-primary"
            type="submit"
          >
            <span>
              Save Changes
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Cancel
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Remove
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`ItemForm renders custom fields submits proper values 1`] = `
<form
  class="ant-form ant-form-horizontal ItemForm-layout"
  id="custom-test-form"
>
  <div
    class="ant-row ant-form-item"
    style="display: none;"
  >
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-input-number"
          >
            <div
              class="ant-input-number-handler-wrap"
            >
              <span
                aria-disabled="true"
                aria-label="Increase Value"
                class="ant-input-number-handler ant-input-number-handler-up ant-input-number-handler-up-disabled"
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="up"
                  class="anticon anticon-up ant-input-number-handler-up-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="up"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                    />
                  </svg>
                </span>
              </span>
              <span
                aria-disabled="true"
                aria-label="Decrease Value"
                class="ant-input-number-handler ant-input-number-handler-down ant-input-number-handler-down-disabled"
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-input-number-handler-down-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="ant-input-number-input-wrap"
            >
              <input
                aria-valuemin="-9007199254740991"
                aria-valuenow="testId"
                autocomplete="off"
                class="ant-input-number-input"
                id="custom-test-form_id"
                min="-9007199254740991"
                role="spinbutton"
                step="1"
                value="testId"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item ant-form-item-has-success"
  >
    <div
      class="ant-col ant-form-item-label ant-col-xs-16 ant-col-sm-8"
    >
      <label
        class="ant-form-item-required"
        for="custom-test-form_numfield"
        title="Number Field"
      >
        Number Field
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <div
            class="ant-input-number"
          >
            <div
              class="ant-input-number-handler-wrap"
            >
              <span
                aria-disabled="false"
                aria-label="Increase Value"
                class="ant-input-number-handler ant-input-number-handler-up "
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="up"
                  class="anticon anticon-up ant-input-number-handler-up-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="up"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                    />
                  </svg>
                </span>
              </span>
              <span
                aria-disabled="false"
                aria-label="Decrease Value"
                class="ant-input-number-handler ant-input-number-handler-down "
                role="button"
                unselectable="unselectable"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-input-number-handler-down-inner"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="ant-input-number-input-wrap"
            >
              <input
                aria-valuemin="-9007199254740991"
                aria-valuenow="3"
                autocomplete="off"
                class="ant-input-number-input"
                id="custom-test-form_numfield"
                min="-9007199254740991"
                role="spinbutton"
                step="1"
                value="3"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item ant-form-item-has-success"
  >
    <div
      class="ant-col ant-form-item-label ant-col-xs-16 ant-col-sm-8"
    >
      <label
        class=""
        for="custom-test-form_text-field"
        title="Text Field"
      >
        Text Field
      </label>
    </div>
    <div
      class="ant-col ant-col-12 ant-form-item-control"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <input
            class="ant-input"
            id="custom-test-form_text-field"
            type="text"
            value="field value"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-control ant-col-xs-16 ant-col-xs-offset-0 ant-col-sm-16 ant-col-sm-offset-8"
    >
      <div
        class="ant-form-item-control-input"
      >
        <div
          class="ant-form-item-control-input-content"
        >
          <button
            class="ant-btn ant-btn-primary"
            type="submit"
          >
            <span>
              Save Changes
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Cancel
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Remove
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;
